azure-mgmt-appservice/src/main/java/com/microsoft/azure/management/appservice/implementation/AppServiceBaseImpl.java

```diff
@@ -436,6 +439,73 @@ public FluentImplT withExistingAppServicePlan(AppServicePlan appServicePlan) {
         return withOperatingSystem(appServicePlanOperatingSystem(appServicePlan));
     }
 
+    public FluentImplT withPublicDockerHubImage(String imageAndTag) {
+        ensureLinuxPlan();
+        cleanUpContainerSettings();
+        if (siteConfig == null) {
+            siteConfig = new SiteConfigResourceInner();
+        }
+        siteConfig.withLinuxFxVersion(String.format("DOCKER|%s", imageAndTag));
+        withAppSetting(SETTING_DOCKER_IMAGE, imageAndTag);
+        return (FluentImplT) this;
+    }
+
+    public FluentImplT withPrivateDockerHubImage(String imageAndTag) {
+        return withPublicDockerHubImage(imageAndTag);
+    }
+
+    public FluentImplT withPrivateRegistryImage(String imageAndTag, String serverUrl) {
+        imageAndTag = smartCompletionPrivateRegistryImage(imageAndTag, serverUrl);
+
+        ensureLinuxPlan();
+        cleanUpContainerSettings();
+        if (siteConfig == null) {
+            siteConfig = new SiteConfigResourceInner();
+        }
+        siteConfig.withLinuxFxVersion(String.format("DOCKER|%s", imageAndTag));
+        withAppSetting(SETTING_DOCKER_IMAGE, imageAndTag);
+        withAppSetting(SETTING_REGISTRY_SERVER, serverUrl);
+        return (FluentImplT) this;
+    }
+
+    public FluentImplT withCredentials(String username, String password) {
+        withAppSetting(SETTING_REGISTRY_USERNAME, username);
+        withAppSetting(SETTING_REGISTRY_PASSWORD, password);
+        return (FluentImplT) this;
+    }
+
+    protected abstract void cleanUpContainerSettings();
+
+    private static String smartCompletionPrivateRegistryImage(String imageAndTag, String serverUrl) {
+        try {
+            URL url = new URL(serverUrl);
+            String registryServer = url.getAuthority();
+            if (!registryServer.isEmpty() && !imageAndTag.trim().startsWith(registryServer)) {
+                String[] segments = imageAndTag.split(Pattern.quote("/"));
+                if (segments.length == 1) {
+                    // it appears that imageAndTag does not contain registry server, add registry server before it.
+                    imageAndTag = String.format("%s/%s", registryServer, imageAndTag);
+                }
+                if (segments.length > 1) {
+                    String segment = segments[0];
+                    if (!segment.isEmpty() && !segment.contains(".") && !segment.contains(":") && !segment.equals(registryServer)) {
```

Comment:
`.` can be in image name, and `:` can not be dealt by docker in docker/kitematic #1840

Reply:
This is just a tool for docker, can be ignored.